causallearn/utils/cit.py

```diff
@@ -15,78 +15,142 @@
-            'mc_fisherz': self.mc_fisherz,
-            'kci': self.kci,
-            'chisq': self.chisq,
-            'gsq': self.gsq
-        }
-
-    def kci(self, X, Y, condition_set):
-        if type(X) == int:
-            X = [X]
-        elif type(X) != list:
-            Y = list(X)
-        if type(Y) == int:
-            Y = [Y]
-        elif type(Y) != list:
-            Y = list(Y)
-        
-        if len(condition_set) == 0:
-            return self.kci_ui.compute_pvalue(self.data[:, X], self.data[:, Y])[0]
-        return self.kci_ci.compute_pvalue(self.data[:, X], self.data[:, Y], self.data[:, list(condition_set)])[0]
-
-    def fisherz(self, X, Y, condition_set):
-        """
-        Perform an independence test using Fisher-Z's test
+            assert self.pvalue_cache['method_name'] == method_name, "CI test method name mismatch." # a loaded cache
+            assert self.pvalue_cache['parameters_hash'] == parameters_hash, "CI test method parameters mismatch."
+
+    def assert_input_data_is_valid(self, allow_nan=False, allow_inf=False):
+        assert allow_nan or not np.isnan(self.data).any(), "Input data contains NaN. Please check."
+        assert allow_inf or not np.isinf(self.data).any(), "Input data contains Inf. Please check."
+
+    def save_to_local_cache(self):
+        if not self.cache_path is None and time.time() - self.last_time_cache_saved > self.SAVE_CACHE_CYCLE_SECONDS:
+            with codecs.open(self.cache_path, 'w') as fout: fout.write(json.dumps(self.pvalue_cache, indent=2))
+            self.last_time_cache_saved = time.time()
+
+    def get_formatted_XYZ_and_cachekey(self, X, Y, condition_set):
+        '''
+        reformat the input X, Y and condition_set to
+            1. convert to built-in types for json serialization
+            2. handle multi-dim unconditional variables (for kernel-based)
+            3. basic check for valid input (X, Y no overlap with condition_set)
+            4. generate unique and hashable cache key
+
+        Parameters
+        ----------
+        X: int, or np.*int*
+        Y: int, or np.*int*
+        condition_set: Iterable<int | np.*int*>
+
+        Returns
+        -------
+        Xs: List<int>, sorted. may swapped with Ys for cache key uniqueness.
+        Ys: List<int>, sorted.
+        condition_set: List<int>
+        cache_key: string. Unique for <X,Y|S> in any input type or order.
+        '''
+        def _stringize(ulist1, ulist2, clist):
+            # ulist1, ulist2, clist: list of ints, sorted.
+            _strlst  = lambda lst: '.'.join(map(str, lst))
+            return f'{_strlst(ulist1)};{_strlst(ulist2)}|{_strlst(clist)}' if len(clist) > 0 else \
+                   f'{_strlst(ulist1)};{_strlst(ulist2)}'
+
+        # every time when cit is called, auto save to local cache.
+        self.save_to_local_cache()
+
+        METHODS_SUPPORTING_MULTIDIM_DATA = ["kci"]
+        if condition_set is None: condition_set = []
```

Comment:
might be cleaner to force condition_set is never None? 

If the user don't want condition_set, just use []?

Reply:
Oh I missed this.

Sometimes users might want to input `FisherZ(X, Y)` to test just unconditional independence.

From the users' end, I feel that this looks better than `FisherZ(X, Y, [])`, or `FisherZ(X, Y, ())` ... (which of course also works).

Another reason is that, I'm not sure whether there exist usages like `FisherZ(X, Y)` in current codes. lol